File: app/src/main/java/com/morrowbone/mafiacards/app/adapter/CreateDeckArrayAdapter.java
package com.morrowbone.mafiacards.app.adapter;

/**
 * Created by morrow on 31.07.2014.
 */

import android.content.Context;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.morrowbone.mafiacards.app.R;
import com.morrowbone.mafiacards.app.activity.CreatorActivity;
import com.morrowbone.mafiacards.app.model.Card;
import com.morrowbone.mafiacards.app.utils.Constants;

import java.util.List;

public class CreateDeckArrayAdapter extends ArrayAdapter<Card> {

    private static final Integer layout_id = R.layout.view_creator_cart;
    private final Typeface mTypeFace;
    private LayoutInflater mInflater;
    private CreatorActivity mContext;
    private static Integer mCardCount = 0;

    public CreateDeckArrayAdapter(Context context, List<Card> values) {
        super(context, layout_id, values);
        mInflater = (LayoutInflater) getContext().getSystemService(
                Context.LAYOUT_INFLATER_SERVICE);
        mContext = (CreatorActivity) context;
        mTypeFace = Typeface.createFromAsset(mContext.getAssets(), Constants.getTypeFacePath());
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final Card item = getItem(position);
        final Holder holder;

        if (convertView == null) {
            convertView = mInflater.inflate(layout_id, parent, false);

            holder = new Holder();
            holder.title = (TextView) convertView
                    .findViewById(R.id.game_title);
            holder.title.setTypeface(mTypeFace);

            holder.image = (ImageView) convertView.findViewById(R.id.game_type_image);

            holder.decrement = convertView.findViewById(R.id.decrement);
            holder.increment = convertView.findViewById(R.id.increment);
            holder.cardCount = (TextView) convertView.findViewById(R.id.card_count);
            holder.cardCount.setTypeface(mTypeFace);
            convertView.setTag(holder);
        } else {
            holder = (Holder) convertView.getTag();
        }

        holder.title.setText(item.getRoleNameStringId());
        holder.image.setImageResource(item.getCartFrontSideImageId());
        holder.image.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

            }
        });
        final Integer[] curCount = {item.getCountInDeck()};

        holder.cardCount.setText(String.valueOf(curCount[0]));
        holder.decrement.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (curCount[0] > 0) {
                    curCount[0]--;
                    holder.cardCount.setText(String.valueOf(curCount[0]));
                    item.setCountInDeck(curCount[0]);

                    mCardCount--;
                    mContext.setCardCount(mCardCount);
                }
            }
        });

        holder.increment.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (curCount[0] < 100) {
                    curCount[0]++;
                    item.setCountInDeck(curCount[0]);
                    holder.cardCount.setText(String.valueOf(curCount[0]));

                    mCardCount++;
                    mContext.setCardCount(mCardCount);
                }
            }
        });

        return convertView;
    }

    /**
     * View holder for the views we need access to
     */
    private static class Holder {
        private TextView title;
        private ImageView image;
        private View increment;
        private View decrement;
        private TextView cardCount;
    }

}

File: app/src/main/java/com/morrowbone/mafiacards/app/adapter/CreateDeckArrayAdapter.kt
package com.morrowbone.mafiacards.app.adapter

/**
 * Created by morrow on 31.07.2014.
 */

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.core.view.isVisible
import com.morrowbone.mafiacards.app.R
import com.morrowbone.mafiacards.app.data.*

class CreateDeckArrayAdapter(
        context: Context,
        values: List<AbstractCard>,
        private val onCardCountChanged: (Int) -> Unit,
        private val editCardCallback: (cardId: String) -> Unit
) : ArrayAdapter<AbstractCard>(context, layout_id, values) {
    private val defaultCards = mutableListOf<DefaultCard>()
    private val userCards = mutableListOf<Card>()
    val deck = Deck(DeckRepository.USER_DECK, CardsSet(defaultCards, userCards))

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        val item = getItem(position)!!
        val itemId = item.getId()
        val holder: Holder

        if (convertView == null) {
            convertView = LayoutInflater.from(parent.context).inflate(layout_id, parent, false)

            holder = Holder()
            holder.title = convertView!!
                    .findViewById<View>(R.id.card_title) as TextView

            holder.image = convertView.findViewById<View>(R.id.game_type_image) as ImageView

            holder.decrement = convertView.findViewById(R.id.decrement)
            holder.increment = convertView.findViewById(R.id.increment)
            holder.cardCount = convertView.findViewById<View>(R.id.card_count) as TextView
            holder.editIcon = convertView.findViewById(R.id.editIcon)
            holder.titleLayout = convertView.findViewById(R.id.card_title_layout)
            convertView.tag = holder
        } else {
            holder = convertView.tag as Holder
        }

        holder.title!!.text = item.getTitle()
        holder.image!!.setImageResource(item.getImageResId())

        var curCount = deck.getCards().count { it.getId() == itemId }

        holder.cardCount!!.text = curCount.toString()
        holder.decrement!!.setOnClickListener {
            if (curCount > 0) {
                curCount--
                holder.cardCount!!.text = curCount.toString()
                mCardCount--
                onCardCountChanged.invoke(mCardCount)
                removeCard(item)
            }
        }

        holder.increment!!.setOnClickListener {
            if (curCount < 100) {
                curCount++
                holder.cardCount!!.text = curCount.toString()

                mCardCount++
                onCardCountChanged.invoke(mCardCount)
                addCard(item)
            }
        }

        val isUserCard = item is Card
        holder.editIcon!!.isVisible = isUserCard
        if (isUserCard) {
            holder.titleLayout!!.setOnClickListener {
                onEditClick(itemId)
            }
            holder.image!!.setOnClickListener { onEditClick(itemId) }
        } else {
            holder.titleLayout!!.setOnClickListener(null)
            holder.image!!.setOnClickListener(null)
        }

        return convertView
    }

    private fun onEditClick(cardId: String) {
        editCardCallback.invoke(cardId)
    }

    private fun addCard(card: AbstractCard) {
        if (card is DefaultCard) {
            defaultCards.add(card)
        } else if (card is Card) {
            userCards.add(card)
        }
    }

    private fun removeCard(card: AbstractCard) {
        if (card is DefaultCard) {
            defaultCards.remove(card)
        } else if (card is Card) {
            userCards.remove(card)
        }
    }

    fun updateDeck(deck: Deck?) {
        if (deck == null) {
            return
        }

        defaultCards.clear()
        userCards.clear()

        defaultCards.addAll(deck.cardsSet.defaultCards)
        userCards.addAll(deck.cardsSet.userCards)
        notifyDataSetChanged()
    }

    /**
     * View holder for the views we need access to
     */
    private class Holder {
        var title: TextView? = null
        var image: ImageView? = null
        var increment: View? = null
        var decrement: View? = null
        var cardCount: TextView? = null
        var editIcon: View? = null
        var titleLayout: ViewGroup? = null
    }

    companion object {
        private val layout_id = R.layout.view_creator_card
        private var mCardCount: Int = 0
    }
}
